### PR TITLE
Raise appropriate Exceptions in place of generic exceptions

### LINKED ISSUE #2167: PlotDataItem.setData() raises generic Exception instead of TypeError.
When using setData() method in the context of a try: except statement (plenty possible application since that's the low level way of redrawing fast) one has to resort to a generic Exception catch which looks nooby and can lead to undetected bugs (just what happened to me). setData() should probably be raising more specific exception types, judging by the Exception messages in place these should probably both be TypeErrors.

### Short description
PlotDataItem.setData() raises generic Exception instead of TypeError.

### Code to reproduce 1
```
import pyqtgraph as pg
plotWidget = pg.plot()
plot_data_item = plotWidget.plot()
plot_data_item.setData(None)
```

### Expected behavior
```
TypeError: Invalid data type <class 'NoneType'>
```

### Real behavior
```
Exception: Invalid data type <class 'NoneType'>
```

### Code to reproduce 2
```
import pyqtgraph as pg
plotWidget = pg.plot()
plot_data_item = plotWidget.plot()
plot_data_item.setData(None,None)
```

### Expected behavior
```
TypeError: When passing two unnamed arguments, both must be a list or array of values. (got <class 'NoneType'>, <class 'NoneType'>)
```

### Real behavior
```
Exception: When passing two unnamed arguments, both must be a list or array of values. (got <class 'NoneType'>, <class 'NoneType'>)
```

### Tested environment(s)
  - python=3.7.6=cpython_h60c2a47_6
  - pyqtgraph=0.11.0=pyh9f0ad1d_0 (this is certainly not the latest but the raise Except statements are there in master still)

PS: There is another 'raise Exception' in PlotDataItem.py relating to wrong dimensioning on numpy arrays. Here some specific class ArrayDimensionError(Exception) could be created or perhaps numpy readily implements some applicable exception type.

### PR DESCRIPTION
**Pull request by Nibba2018**
Fixes #2167 
Raise `TypeError` in place of generic `Exception` for `PlotDataItem.setData()`.
Introduced and raised a new exception named `ArrayDimensionError` when numpy arrays have an Invalid shape.

### Other Tasks 

<details>
  <summary>Raise `TypeError` in place of generic `Exception` for `PlotDataItem.setData()`</summary>

### Files that need updates
    
Confirm the following files have been either updated or there has been a determination that no update is needed.

- [X] `README.md`
- [X] `setup.py`
- [X] `tox.ini`
- [X] `.github/workflows/main.yml` and associated `requirements.txt` and conda `environemt.yml` files
- [X] `pyproject.toml`
- [X] `binder/requirements.txt`

</details>

<details>
    <summary>Pre-Release Checklist</summary>

### Pre Release Checklist

- [ ] Update version info in `__init__.py`
- [ ] Update `CHANGELOG` primarily using contents from automated changelog generation in GitHub release page
- [ ] Have git tag in the format of pyqtgraph-<version>

</details>

<details>
  <summary>Post-Release Checklist</summary>

### Steps To Complete

- [ ] Append `.dev0` to `__version__` in `__init__.py`
- [ ] Announce on mail list
- [ ] Announce on Twitter

</details>

### CODE REVIEW

File: pyqtgraph/graphicsItems/PlotDataItem.py [hunk 1]
```diff
@@ -1214,7 +1215,7 @@ def dataType(obj):
             elif obj.ndim == 2 and obj.dtype.names is None and obj.shape[1] == 2:
                 return 'Nx2array'
             else:
-                raise Exception('array shape must be (N,) or (N,2); got %s instead' % str(obj.shape))
+                raise ArrayDimensionError('array shape must be (N,) or (N,2); got %s instead' % str(obj.shape))
```
Comment (j9ac9k) [inline pyqtgraph/graphicsItems/PlotDataItem.py:1218]: I do think it's a good idea to introduce more specific exceptions for this sort of thing, but I think for this purpose, we should probably stick with `ValueError` which I believe is what numpy raises when you get a dimension mismatch related exceptions.

**Comment by j9ac9k** [comment]
Hi @Nibba2018 

Thanks for the PR! 

I made a comment, while I think it's a good idea to have more specific exceptions, this is something that users of pyqtgraph aren't expecting, so it would create some problems for folks trying to catch this type of exception (they wouldn't know where to import it from).

I think what would be best to stick w/ numpy's convention, where I believe they raise a `ValueError` for this type of issue.

File: pyqtgraph/graphicsItems/PlotDataItem.py (hunk 1 above)
Comment (Nibba2018) [reply pyqtgraph/graphicsItems/PlotDataItem.py:1218]: Thank you for the review @j9ac9k , will update it to `ValueError`.

File: pyqtgraph/graphicsItems/PlotDataItem.py [hunk 2]
```diff
@@ -1214,7 +1214,7 @@ def dataType(obj):
             elif obj.ndim == 2 and obj.dtype.names is None and obj.shape[1] == 2:
                 return 'Nx2array'
             else:
-                raise Exception('array shape must be (N,) or (N,2); got %s instead' % str(obj.shape))
+                raise TypeError('array shape must be (N,) or (N,2); got %s instead' % str(obj.shape))
```
Comment (j9ac9k) [inline pyqtgraph/graphicsItems/PlotDataItem.py:1217]: Hi, this one should be a `ValueError` not a `TypeError` 👍🏻

**Comment by j9ac9k** [comment]
diff doesn't match the comment/message, looks like it was replaced with a TypeError ;)

File: pyqtgraph/graphicsItems/PlotDataItem.py (hunk 2 above)
Comment (Nibba2018) [reply pyqtgraph/graphicsItems/PlotDataItem.py:1217]: OMG sorry about that :rofl:

**Comment by j9ac9k** [comment]
LGTM thanks for the contribution @Nibba2018 !

**Pull request merged**